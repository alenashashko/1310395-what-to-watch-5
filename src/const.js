export const MillisecondsInTimePeriod = {
  SECOND: 1000,
  MINUTE: 60000,
  HOUR: 3600000,
};
