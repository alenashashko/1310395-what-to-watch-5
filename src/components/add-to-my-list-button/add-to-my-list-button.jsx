import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import proptypes from '../../type';
import {getAuthorizationStatus} from '../../store/selectors';
import {MovieChangeStatus, AppRoute, AuthorizationStatus} from '../../const';
import {changeFavoriteMovieByID} from '../../store/api-actions';
import {redirectToRoute} from '../../store/actions';

const AddToMyListButton = (props) => {
  const {id, isMovieFavorite, changeFavoriteMovieByIDAction} = props;

  const handleAddToFavoriteButtonClick = () => {
    const {authorizationStatus, redirectToAuthPage} = props;

    if (authorizationStatus !== AuthorizationStatus.AUTH) {
      redirectToAuthPage();

      return false;
    }

    return true;
  };

  return (
    isMovieFavorite
      ? <button
        onClick={
          () => {
            const isLoggedIn = handleAddToFavoriteButtonClick();

            if (isLoggedIn) {
              changeFavoriteMovieByIDAction(id, MovieChangeStatus.DELETE_FROM_FAVORITE);
            }
          }
        }
        className="btn btn--list movie-card__button"
        type="button">
        <svg viewBox="0 0 18 14" width="18" height="14">
          <use xlinkHref="#in-list"></use>
        </svg>
        <span>My list</span>
      </button>
      : <button
        onClick={
          () => {
            const isLoggedIn = handleAddToFavoriteButtonClick();

            if (isLoggedIn) {
              changeFavoriteMovieByIDAction(id, MovieChangeStatus.ADD_TO_FAVORITE);
            }
          }
        }
        className="btn btn--list movie-card__button"
        type="button">
        <svg viewBox="0 0 19 20" width="19" height="20">
          <use xlinkHref="#add"></use>
        </svg>
        <span>My list</span>
      </button>
  );
};

AddToMyListButton.propTypes = {
  id: proptypes.id,
  isMovieFavorite: PropTypes.bool.isRequired,
  changeFavoriteMovieByIDAction: PropTypes.func.isRequired,
  redirectToAuthPage: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  changeFavoriteMovieByIDAction(id, status) {
    dispatch(changeFavoriteMovieByID(id, status));
  },
  redirectToAuthPage() {
    dispatch(redirectToRoute(AppRoute.LOGIN));
  }
});

export {AddToMyListButton};
export default connect(mapStateToProps, mapDispatchToProps)(AddToMyListButton);
