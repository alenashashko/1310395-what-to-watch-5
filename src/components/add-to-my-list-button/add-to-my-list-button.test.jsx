import React from 'react';
import renderer from 'react-test-renderer';

import {AddToMyListButton} from './add-to-my-list-button';
import {AuthorizationStatus} from '../../const';

const noop = () => {};

describe(`Should AddToMyListButton render correctly`, () => {
  it(`With favorite movie when user authorised`, () => {
    const tree = renderer
      .create(
          <AddToMyListButton
            id="1"
            isMovieFavorite={true}
            changeFavoriteMovieByIDAction={noop}
            redirectToAuthPage={noop}
            authorizationStatus={AuthorizationStatus.AUTH}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With favorite movie when user unauthorised`, () => {
    const tree = renderer
      .create(
          <AddToMyListButton
            id="1"
            isMovieFavorite={true}
            changeFavoriteMovieByIDAction={noop}
            redirectToAuthPage={noop}
            authorizationStatus={AuthorizationStatus.NO_AUTH}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With not favorite movie when user authorised`, () => {
    const tree = renderer
      .create(
          <AddToMyListButton
            id="1"
            isMovieFavorite={false}
            changeFavoriteMovieByIDAction={noop}
            redirectToAuthPage={noop}
            authorizationStatus={AuthorizationStatus.AUTH}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With not favorite movie when user unauthorised`, () => {
    const tree = renderer
      .create(
          <AddToMyListButton
            id="1"
            isMovieFavorite={false}
            changeFavoriteMovieByIDAction={noop}
            redirectToAuthPage={noop}
            authorizationStatus={AuthorizationStatus.NO_AUTH}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
