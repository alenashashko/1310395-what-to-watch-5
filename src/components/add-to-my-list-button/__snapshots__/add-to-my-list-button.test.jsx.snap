// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should AddToMyListButton render correctly With favorite movie and AUTH authorization status 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="14"
    viewBox="0 0 18 14"
    width="18"
  >
    <use
      xlinkHref="#in-list"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;

exports[`Should AddToMyListButton render correctly With favorite movie and NO_AUTH authorization status 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="14"
    viewBox="0 0 18 14"
    width="18"
  >
    <use
      xlinkHref="#in-list"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;

exports[`Should AddToMyListButton render correctly With not favorite movie and AUTH authorization status 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="20"
    viewBox="0 0 19 20"
    width="19"
  >
    <use
      xlinkHref="#add"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;

exports[`Should AddToMyListButton render correctly With not favorite movie and NO_AUTH authorization status 1`] = `
<button
  className="btn btn--list movie-card__button"
  onClick={[Function]}
  type="button"
>
  <svg
    height="20"
    viewBox="0 0 19 20"
    width="19"
  >
    <use
      xlinkHref="#add"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;
